File: java/206.reverse-linked-list.java
/*
 * @lc app=leetcode id=206 lang=java
 *
 * [206] Reverse Linked List
 */
class Solution206 {

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// @lc code=start
// class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null){
            return head;
        }
        ListNode pre = head;
        ListNode cur = head.next;
        ListNode nxt;
        while(cur!=null){
            nxt = cur.next;
            cur.next = pre;

            pre = cur;
            cur = nxt;
        }
        head.next = null;
        return pre;
    }
}
// @lc code=end

File: go/leetcode/206.reverse-linked-list.go
package leetcode

/*
 * @lc app=leetcode id=206 lang=golang
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	var pre *ListNode
	cur := head
	for cur != nil {
		// cur.Next, pre, cur = pre, cur, cur.Next
		pre, cur, cur.Next = cur, cur.Next, pre
	}
	return pre

}

// @lc code=end
